keps/sig-node/1898-hardened-exec/README.md

```diff
@@ -0,0 +1,531 @@
+authenticated request by:
+
+- Eliminating the ability to control an executed command with a URL parameter.
+- Limiting the ability to execute commands with a GET method, without breaking websocket connections.
+
+### Non-Goals
+
+- Protecting proxy requests
+- Maintaining backwards compatibility of the Kubelet API
+- Maintaining long-term backwards compatibility of risky exec requests
+
+## Proposal
+
+### Kubelet API
+
+The changes to the Kubelet API are:
+
+1. Remove the `/run` endpoint, which is unused in core Kubernetes.
+2. Require a `POST` request for the `/exec`, `/attach` and `/portForward` endpoints.
+3. Remove the `/{UID}/` versions of the endpoints.
+4. Require the exec/attach/port-forward options and pod reference to be provided in the request body.
+
+**Note on backwards compatibility:** This proposal assumes that the only client that should be
+talking to these endpoints on the Kubelet is the kube-apiserver. Therefore, while the changes must
+support API server version skew, endpoints and request types that are unused by the kube-apiserver
+can be removed immediately.
+
+#### 1. Remove the `/run` endpoint
+
+The `run` endpoint provides the option to run a command in a container with a synchronous
+request. This endpoint is not used by any Kubernetes components, and should be removed to reduce the
+attack surface.
+
+**Risk:** Although this endpoint is not exposed in the Kubernetes API, it is still reachable via a
+proxy request to a node, or by connecting to the Kubelet directly. If desired, we could tie the
+removal to the deprecation feature gate discussed
+[below](#1-require-options-to-be-included-in-the-post-request-body-for-exec-requests).
+
+#### 2. Require a `POST` request to the streaming endpoints
+
+Exec, attach, and port-forward (and run) all respond to either GET or POST requests. Currently the
+kube-apiserver uses the same request method when calling the Kubelet as the incoming client request
+(e.g. a GET exec request to the apiserver results in a GET exec request to the kubelet). The API
+server will need to be updated to always POST to these endpoints, regardless of the incoming client
+request.
+
+Since we do not support Kubelets newer than the kube-apiserver, we can safely remove support for GET
+requests to these endpoints in the same release that the apiserver moves to unconditional POST.
+
+Although attach & port-forward do not have the same risks of arbitrary code execution as exec, they
+share a lot of the same code and should match the exec logic to reduce complexity.
+
+#### 3. Remove the `UID` versions of the endpoints
+
+The exec, attach, and port-forward handlers all support 2 URL path formats:
+
+1. `/exec/{podNamespace}/{podID}/{containerName}`
+2. `/exec/{podNamespace}/{podID}/{uid}/{containerName}`
+
+The second path format includes the pod's UID, and is never called by the kube-apiserver. We should
+remove these endpoints as they're unused, to reduce code complexity.
```

Comment:
Aside: if you wanted to preserve the UID checking behavior, it might be feasible to make up a conditional header `If-UID:` and fail requests where that's set and not the same as the running Pod's UID.

Reply:
I don't think it's used anywhere. It would be easy enough to add this approach later if a use case arose.